File: StackAndQueue/CelebrityProblem.java
package StackAndQueue;

import java.util.ArrayDeque;
import java.util.Deque;

public class CelebrityProblem {

    // User function Template for Java

    // Function to find if there is a celebrity in the party or not.
    int findCelebrity(int[][] partyMatrix, int partySize) {
        // Use a stack to track potential celebrities
        Deque<Integer> potentialCelebrities = new ArrayDeque<>();

        // Populate the stack with all party attendees
        for (int i = 0; i < partySize; i++) {
            potentialCelebrities.push(i);
        }

        // While there is more than one potential celebrity remaining
        while (potentialCelebrities.size() > 1) {
            int personA = potentialCelebrities.pop();
            int personB = potentialCelebrities.pop();
            // If person A knows person B, person A cannot be a celebrity, so person B is
            // pushed back onto the stack
            if (knows(personA, personB, partyMatrix)) {
                potentialCelebrities.push(personB);
            } else {
                // If person A does not know person B, person B cannot be a celebrity, so person
                // A is pushed back onto the stack
                potentialCelebrities.push(personA);
            }
        }
        // The remaining person in the stack is the potential celebrity
        int potentialCelebrity = potentialCelebrities.peek();

        // Verify the potential celebrity
        int zeroCount = 0;
        // Check the row of the potential celebrity for zeros
        for (int i = 0; i < partySize; i++) {
            if (partyMatrix[potentialCelebrity][i] == 0) {
                zeroCount++;
            }
        }
        // If there are non-zero entries in the row, the potential celebrity is not a
        // celebrity
        if (zeroCount != partySize) {
            return -1;
        }

        int oneCount = 0;
        // Check the column of the potential celebrity for ones
        for (int i = 0; i < partySize; i++) {
            if (partyMatrix[i][potentialCelebrity] == 1) {
                oneCount++;
            }
        }
        // If there are not exactly n-1 ones in the column, the potential celebrity is
        // not a celebrity
        if (oneCount != partySize - 1) {
            return -1;
        }

        return potentialCelebrity;
    }

    boolean knows(int a, int b, int[][] M) {
        return M[a][b] == 1;
    }
}

File: StackAndQueue/Medium/CelebrityProblem.py
#User function Template for python3

class Solution:
    
    #Function to find if there is a celebrity in the party or not.
    def celebrity(self, M, n):
        stack=[]
        
        for i in range(n):
            stack.append(i)
        
        
        while len(stack)!=1:
            a= stack.pop()
            b=stack.pop()
            if self.knows(a,b,M):
                stack.append(b)
            else:
                stack.append(a)
        candidate= stack[-1]
        
        
        #Verify now
        row_check=False
        zeroCount=0
        for i in range(n):
            if M[candidate][i]==0:
                zeroCount+=1
        if zeroCount!=n:
            return -1
        
        one_count=0
        col_check=False
        for i in range(n):
            if M[i][candidate]==1:
                one_count+=1
        if one_count!=n-1:
            return -1
        
        return candidate
    
    def knows(self,a,b,M):
        if M[a][b]==1:
            return True
        else:
            return False
            

